packages/flutter/lib/src/widgets/app.dart

```diff
@@ -677,33 +749,80 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
+    // Attempt to use localeListResolutionCallback.
+    if (widget.localeListResolutionCallback != null) {
+      final Locale locale = widget.localeListResolutionCallback(preferredLocales, widget.supportedLocales);
+      if (locale != null)
+        return locale;
+    }
+    // localeListResolutionCallback failed, falling back to localeResolutionCallback.
     if (widget.localeResolutionCallback != null) {
-      final Locale locale = widget.localeResolutionCallback(newLocale, widget.supportedLocales);
+      final Locale locale = widget.localeResolutionCallback(preferredLocales.first, widget.supportedLocales);
       if (locale != null)
         return locale;
     }
+    // Both callbacks failed, falling back to default algorithm.
+    return _fallbackLocaleResolution(preferredLocales, supportedLocales);
+  }
+
+  static Locale _fallbackLocaleResolution(List<Locale> preferredLocales, Iterable<Locale> supportedLocales) {
     // newLocale can be null when called before the platform has had a chance to
     // initialize the locales. We default to the first supported locale.
-    if (newLocale == null) {
+    if (preferredLocales == null || preferredLocales.isEmpty) {
       return supportedLocales.first;
     }
-
-    Locale matchesLanguageCode;
+    // Hash the supported locales because apps can support many locales and would
+    // be expensive to search through them many times.
+    Map<int, Locale> allSupportedLocales = HashMap<int, Locale>();
+    Map<int, Locale> languageAndCountryLocales = HashMap<int, Locale>();
+    Map<int, Locale> languageAndScriptLocales = HashMap<int, Locale>();
+    Map<int, Locale> languageLocales = HashMap<int, Locale>();
     for (Locale locale in supportedLocales) {
-      if (locale == newLocale)
-        return newLocale;
-      if (locale.languageCode == newLocale.languageCode)
-        matchesLanguageCode ??= locale;
+      allSupportedLocales[ui.hashValues(locale.languageCode.hashCode, locale.scriptCode.hashCode, locale.countryCode.hashCode)] ??= locale;
+      languageAndCountryLocales[ui.hashValues(locale.languageCode.hashCode, locale.countryCode)] ??= locale;
+      languageAndScriptLocales[ui.hashValues(locale.languageCode.hashCode, locale.scriptCode.hashCode)] ??= locale;
+      languageLocales[locale.languageCode.hashCode] ??= locale;
+    }
+    // Loop over user's preferred locales
+    Locale matchesLanguageCode;
+    for (int localeIndex = 0; localeIndex < preferredLocales.length; localeIndex++) {
+      Locale userLocale = preferredLocales[localeIndex];
+      // Look for perfect match.
+      if (allSupportedLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode, userLocale.countryCode.hashCode))) {
+        return userLocale;
+      }
+      // Wait until next userLocale to see if there is a high accuracy match before
+      // returning a language-only match.
```

Comment:
i don't understand this comment

Reply:
Since languageCode only matches are possibly low quality, we don't return it instantly when we find such a match. We check to see if the next preferred locale in the list has a high accuracy match, and only return the languageCode only match when a higher accuracy match in the next preferred locale cannot be found.

Reply:
The way this works is it caches the languageCode-only match and returns it in the next loop iteration as appropriate.